src/backend/FramebufferBackend.cc

```diff
@@ -0,0 +1,200 @@
+        if (ioctl(fbfd, FBIOPUT_VSCREENINFO, &this->origVinfo)) {
+            Nan::ThrowError("Error restoring framebuffer state");
+        }
+        fbp = nullptr;
+    }
+    if (this->fbfd > 0) {
+        close(this->fbfd);
+        this->fbfd = 0;
+    }
+
+    if (cairo_surface_status(this->screenSurface) == CAIRO_STATUS_SUCCESS) {
+        cairo_surface_destroy(this->screenSurface);
+    }
+
+    if (this->bbp != nullptr) {
+        free(this->bbp);
+    }
+    if (cairo_surface_status(this->bufferSurface) == CAIRO_STATUS_SUCCESS) {
+        cairo_surface_destroy(this->bufferSurface);
+    }
+    this->surface = nullptr;
+}
+
+cairo_surface_t* FramebufferBackend::recreateSurface()
+{
+    this->destroySurface();
+    return this->createSurface();
+}
+
+cairo_surface_t* FramebufferBackend::createSurface()
+{
+    // Open the framebuffer device.
+    this->fbfd = open(this->fbdev.c_str(), O_RDWR);
+    if (fbfd < 0) {
+        Nan::ThrowError("Error opening framebuffer device");
+        return nullptr;
+    }
+
+    struct fb_var_screeninfo vinfo;
+    struct fb_fix_screeninfo finfo;
+
+    // Get the screen info.
+    if (ioctl(this->fbfd, FBIOGET_VSCREENINFO, &vinfo)) {
+        this->destroySurface();
+        Nan::ThrowError("Error retrieving data from framebuffer");
+        return nullptr;
+    }
+
+    // Store the original screen info for restoration later.
+    memcpy(&this->origVinfo, &vinfo, sizeof(struct fb_var_screeninfo));
+
+    // Setup the number of bits per pixel.
+    vinfo.bits_per_pixel = 8;
+    if (ioctl(this->fbfd, FBIOPUT_VSCREENINFO, &vinfo)) {
+        this->destroySurface();
+        Nan::ThrowError("Error sending data to framebuffer");
+        return nullptr;
+    }
+
+    // Get the length of the screen buffer in bytes.
+    if (ioctl(this->fbfd, FBIOGET_FSCREENINFO, &finfo)) {
+        this->destroySurface();
+        Nan::ThrowError("Error retrieving data from framebuffer");
+        return nullptr;
+    }
+    this->screenSize = finfo.smem_len;
+
+    // Memory map the framebuffer.
+    this->fbp = (char*)mmap(0,
+                            this->screenSize,
+                            PROT_READ | PROT_WRITE,
+                            MAP_SHARED,
+                            this->fbfd,
+                            0);
+
+    if (this->fbp == MAP_FAILED) {
+        this->destroySurface();
+        Nan::ThrowError("Error during memory mapping");
+        return nullptr;
+    }
+
+    // Create a cairo surface from the memory mapped buffer.
+    screenSurface = cairo_image_surface_create_for_data(
+            (unsigned char *)this->fbp,
+            CAIRO_FORMAT_RGB16_565,
+            vinfo.xres, vinfo.yres,
+            cairo_format_stride_for_width(CAIRO_FORMAT_RGB16_565, vinfo.xres));
+
+    // Store the width and height to be used further up the stack.
+    this->width = vinfo.xres;
+    this->height = vinfo.yres;
+
+    // If double buffering is enabled then allocate an in memory buffer.
```

Comment:
Instead of creating a buffer in memory and copy it, it's more performant to create a single buffer in VRAM with doubled height and just set the offset were it should be readed for each frame, or if that's not possible, at least copy copy the data between the two VRAM regions instead of from the main RAM.

Reply:
That is an interesting idea but I'm not sure how to tell a framebuffer to change where it's looking

Reply:
In the framebuffer docs there's a reference to how to set an offset where it should start reading.